resources/qdrant/src/lib.rs

```diff
@@ -41,62 +34,89 @@ impl Qdrant {
+}
 
-    fn config(&self) -> &Self::Config {
-        self
-    }
+#[async_trait]
+impl IntoResourceInput for Qdrant {
+    type Input = MaybeRequest;
+    // The response can be a provisioned container, depending on local/deployment and config.
+    type Output = Wrapper;
 
-    async fn output(
-        self,
-        factory: &mut dyn Factory,
-    ) -> Result<Self::Output, shuttle_service::Error> {
+    async fn into_resource_input(self, factory: &ResourceFactory) -> Result<Self::Input, Error> {
         let md = factory.get_metadata();
         match md.env {
             Environment::Deployment => match self.cloud_url {
-                Some(cloud_url) => Ok(QdrantClientConfigWrap {
-                    url: cloud_url,
-                    api_key: self.api_key,
-                }),
-                None => Err(ShuttleError::Custom(CustomError::msg(
+                Some(cloud_url) => Ok(MaybeRequest::NotRequest(
+                    QdrantClientConfigWrap {
+                        url: cloud_url,
+                        api_key: self.api_key,
+                    }),
+                ),
+                None => Err(Error::Custom(CustomError::msg(
                     "missing `cloud_url` parameter",
                 ))),
             },
             Environment::Local => match self.local_url {
-                Some(local_url) => Ok(QdrantClientConfigWrap {
-                    url: local_url,
-                    api_key: self.api_key,
-                }),
-                None => {
-                    let port = factory
-                        .get_container(ContainerRequest {
+                Some(local_url) => Ok(MaybeRequest::NotRequest(
+                    QdrantClientConfigWrap {
+                        url: local_url,
+                        api_key: self.api_key,
+                    }),
+                ),
+                None => Ok(MaybeRequest::Request(
+                    ProvisionResourceRequest::new(
+                        Type::Container,
+                        serde_json::to_value(&ContainerRequest {
                             project_name: md.project_name,
                             container_type: "qdrant".to_string(),
-                            image: "docker.io/qdrant/qdrant:latest".to_string(),
+                            image: "docker.io/qdrant/qdrant:v1.7.4".to_string(),
                             port: "6334/tcp".to_string(),
                             env: vec![],
                         })
-                        .await?
-                        .host_port;
-                    let url = format!("http://localhost:{port}");
-
-                    Ok(QdrantClientConfigWrap { url, api_key: None })
-                }
+                        .unwrap(),
+                        serde_json::Value::Null,
+                    )
+                )),
             },
         }
     }
 }
 
+#[derive(Serialize, Deserialize)]
+#[serde(untagged)]
+pub enum Wrapper {
```

Comment:
nit: would be better to name it `OutputWrapper`?

Reply:
It's essentially a private type that no one will look at. Does not really matter.